Exception handling in sys.process

som-snytt opened the pull request:
Handle bad process start and lazily
    
    Improve handling of failure of a process
    to start. In a `CompoundProcess`, catch
    exceptions and set a special exit value.
    For `lazyLines`, also supply a value so
    that requesting thread doesn't hang.
    
    It would be nicer to stow the exception
    and rethrow instead of using special
    integers.
    
    Test for stale thread.

Fixes https://github.com/scala/scala-dev/issues/566

som-snytt committed:
Spawn has a name

PipeThreads also have names.

som-snytt committed:
Handle bad process start and lazily

Improve handling of failure of a process
to start. In a `CompoundProcess`, catch
exceptions and set a special exit value.
For `lazyLines`, also supply a value so
that requesting thread doesn't hang.

It would be nicer to stow the exception
and rethrow instead of using special
integers.

Test for stale thread.

NthPortal commented:
@som-snytt I am flattered that you think I know anything about this

som-snytt committed:
Put output dir on class path for test

som-snytt commented:
@NthPortal A PR is the best place to share our lack of knowledge.

SethTisue commented:
https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-windows/753/

som-snytt commented on test/files/run/t6488.scala:
```diff
@@ -39,9 +39,10 @@ object Test {
+    val command = s"${f.getAbsolutePath} -classpath ${javaClassPath}${pathSeparator}${outdir} Test data"
     Try {
-      Process(command) run ProcessLogger { (s: String) =>
+      Process(command.split(" ").toSeq) run ProcessLogger { (s: String) =>
```
This lazy part assumes no spaces in paths, which can fail on windows.

SethTisue commented:
the Jenkins run is green. let's see if anyone takes an interest in reviewing this. I don't know this code at all

SethTisue commented:
wow, really?

Adriaan and I had given up at this point on the idea of making more modules for 2.13, but that was because it looked like it wasn't going to happen unless someone from the core team did it.

But with 2.13.0-RC1 now pushed out a bit, if you want to go ahead and do it, I'd be in favor. Obviously you'd have our help and support, both with the splitting (which I think would be a pretty straightforward copy-and-paste job), and with helping to drum out additional help from the community once the module has been made.

@adriaanm wdyt? too late, or not?

SethTisue commented:
context, for those just joining us:

* https://contributors.scala-lang.org/t/modularization-of-scala-2-13-standard-library/635
* #5677, #5830, https://github.com/scala/scala/pull/5830/commits/f540fa66bb817a3c5ae0834283fc339680b5989b

SethTisue commented:
looking over those other tickets, I don't see any explicit discussion of `scala.sys.process` being its own module, separate from the rest of `scala.sys`. but actually, I was already tentatively assuming that, as was @som-snytt, apparently. `sys.process` is clearly a coherent unit, while the rest of `scala.sys` is a grab bag. so I think making a scala-process module and leaving the rest of `scala.sys` alone for now would make sense for 2.13.

som-snytt commented:
The other module could be called "better-properties".

lrytz commented on test/files/run/t6488.scala:
```diff
@@ -39,9 +39,10 @@ object Test {
     val count   = new AtomicInteger
     def counted = count.get
-    val command = s"${f.getAbsolutePath} -classpath ${javaClassPath} Test data"
+    val outdir  = s"$userDir/test/files/run/t6488-run.obj"
```
would `sys.props("partest.output")` work?

som-snytt replied:
oh, maybe that.

som-snytt replied:
just for the dirname.

lrytz commented on test/files/run/t6488.scala:
```diff
@@ -39,9 +39,10 @@ object Test {
     def counted = count.get
-    val command = s"${f.getAbsolutePath} -classpath ${javaClassPath} Test data"
+    val outdir  = s"$userDir/test/files/run/t6488-run.obj"
+    val command = s"${f.getAbsolutePath} -classpath ${javaClassPath}${pathSeparator}${outdir} Test data"
```
Why do we need to manually add the outdir now, but it was not necessary before?

som-snytt replied:
No idea. I forgot my Indiana Jones archaeology hat. I could investigate when the windows build started failing.

lrytz replied:
Ah I didn't get the sequence of events. The broken test on windows made you discover the underlying bug. After fixing that, you still fix the test.

Though on non-windows, the manual outdir is not necessary right? Maybe there's a bug in `scala.util.Properties.javaClassPath`? Maybe the test should run in a forked vm?

som-snytt replied:
Actually, I still don't know. I'd assumed it was the `/bin/ls` fallback exec that worked on linux, but no, it should have run java with a `no such class` output. I recall that windows will block if i/o streams aren't consumed, was that the platform difference?

lrytz replied:
Guess what `println(f)` shows? `C:\Program Files\Java\jdk1.8.0_152\jre\bin\java-rmi.exe` :-) Anyway, [that's not the issue](https://stackoverflow.com/questions/25055466/what-is-java-rmi-exe).

The test out dir is missing from the classpath indeed, on linux and windows, so the test never reaches the `data()` part, but that doesn't really matter, as there just needs to be *some* output, and the process produces `Error: Could not find or load main class Test`.

I still don't see why it would hang on windows not on linux, maybe you do?

som-snytt replied:
Now that the Red Sox have clinched it, I'll find time to try it out on Windows. I bought a replacement Windows box a year ago to use for work, but I dread turning it on. I did wonder what exe the test was finding.

My previous comment about windows process blocking if out or err is not consumed is due to this happening to me in a unit test in 2003. That might have been blocking due to hitting a buffer limit, though.

It would be nice to have a `words""` interpolator that uses `CommandLineParser` to split parts and takes args whole to render a `Seq[String]`.

som-snytt replied:
retronym mentioned caveats at the commit to run tests in-process: https://github.com/scala/scala/commit/3b935176bcd0ec22bbbe8317d0dce3b63a88d8bc

(I missed the change at the old repo, so by now, I haven't yet thought about all the implications for tests.)

FSR, listing the obj directory fails when running in-process, after working around "partest.output" not set the same when in-process. However, forcing the test to fork makes it work.

It would be nicer if env was same for in/out-of-process. I'll try to get back to the mystery of why `.obj` is not listable, since compiler output is the same, it's not a virtual dir like in REPL, right?

lrytz reviewed (commented):
running out of steam reviewing this, so much indirection in that module...

SethTisue commented:
re: the question of spinning off the module for 2.13, we discussed it internally and the conclusion was "sorry, feature freeze". there's a lot more to say than that, I have a bunch of notes on it, I'll write more soon, stand by.

(@som-snytt, you might consider whether you're interested in working on making pulling libraries into the REPL easier, a la Ammonite?)

som-snytt commented:
@lrytz Thanks, reviewing is thankless. We discussed this at the office a propos this pr. I spent some time with this over the weekend, because I'd already done the crossword. This code has limited shelf-life, so no point spending time on it, but it was also my fault for approving the previous change to catch exceptions.

som-snytt commented:
@SethTisue no problem, it was just a thought so @lrytz wouldn't spend time reviewing it. He's too nice.

I have a jline3 branch to get back to, then I'll try more ammonization of REPL.

lrytz commented:
@som-snytt how confident are you in the fix? Sorry, hard question :-)

som-snytt commented:
@lrytz I really like the part where the threads get names you can read in a thread dump.

This really just adjusts the previous fix for catching failure on start, by moving the catch up to compound level. Otherwise an i/o thread is started after the failure. The other fix is to the previous lazy lines business, to avoid hanging after failure. So both changes are confident improvements.

som-snytt commented:
@adriaanm @SethTisue for the purposes of semantics, modularization is not a feature, so a feature freeze would not apply. Did you notice David Freese in the world series? My joke was the cold should not bother him, because he is Mr Freese.

lrytz commented on src/library/scala/sys/process/ProcessBuilderImpl.scala:
```diff
@@ -72,19 +73,10 @@ private[process] trait ProcessBuilderImpl {
-             | _: NullPointerException
-             | _: SecurityException
-             | _: UnsupportedOperationException
-        => return FailedProcess
```
It seems `FailedProcess` is now unused.

som-snytt replied:
Thought it might be helpful to have a `FailedProcess` that takes an exception, but you'd need a failed version of each "process" type. I didn't add an exception field either, it still just fails with a negative exit code.

som-snytt commented on test/junit/scala/tools/testing/AssertUtil.scala:
```diff
@@ -97,4 +97,30 @@ object AssertUtil {
+  def assertZeroNetThreads(body: => Unit): Unit = {
+    val beforeCount = Thread.activeCount
+    val beforeThreads = new Array[Thread](beforeCount)
+    assertEquals("Spurious early thread creation.", beforeCount, Thread.enumerate(beforeThreads))
```
Might as well "wait for quiescence" here, too, as it doesn't hurt anything. But this should be done in a new ThreadGroup for isolation, since partest tries to run multiple tests in-process now, IIUC.

lrytz replied:
Or just run this test in a forked jvm, we do it for some others. Just add a `.javaopts` file (https://github.com/scala/scala/blob/2.13.x/test/files/run/t2318.javaopts).

som-snytt replied:
I went with thread isolation.

som-snytt committed:
Zero net thread check in subgroup

SethTisue commented:
> @adriaanm @SethTisue for the purposes of semantics, modularization is not a feature, so a feature freeze would not apply.

I tried that argument, it didn't work on Mr. Adriaan Freeze, aka the Snow Miser:

<img width="467" alt="screen shot 2018-11-03 at 6 37 04 pm" src="https://user-images.githubusercontent.com/161079/47955371-a4a2f980-df97-11e8-81d8-3a48e59aaba3.png">

som-snytt committed:
Clean up t6488 and run forked mode

som-snytt commented:
I think he spells it "Vries". BTW, that's not to say there's not a good argument for being conservative with the change; I just think it needs a different legal justification. Maybe it's the principle that thankless work should be deferred until tomorrow. Or nothing good happens after midnight. And if it ain't broke.

som-snytt committed:
Test and fix for sequential processes

Add contributed hello, world test,
tweaked, with fix, which partially reverts
a previous change that broke coordination
between sequential process and its pipe.

som-snytt commented:
It's not even April Fool's:
```
[test] !! 14 - presentation/doc                          [output differs][duration 11.56s]
[test] % diff /home/jenkins/workspace/scala-2.13.x-validate-main@2/test/files/presentation/doc.check /home/jenkins/workspace/scala-2.13.x-validate-main@2/test/files/presentation/doc-presentation.log
[test] @@ -1,2 +1,4 @@
[test]  warning: there was one deprecation warning (since 2.13.0); re-run with -deprecation for details
[test]  reload: Base.scala, Class.scala, Derived.scala
[test] +Unexpected Derived class comment:None
[test] +Unexpected foo method comment:None
[test] 
[test] ##### Log file '/home/jenkins/workspace/scala-2.13.x-validate-main@2/test/files/presentation/doc-presentation.log' from failed test #####
[test] 
[test] warning: there was one deprecation warning (since 2.13.0); re-run with -deprecation for details
[test] reload: Base.scala, Class.scala, Derived.scala
[test] Unexpected Derived class comment:None
[test] Unexpected foo method comment:None
```

som-snytt commented:
/rebuild

som-snytt commented:
Full disclosure, I still miss [Tor de Vries](https://tordevries.net/portfolio/) whom I knew on the East coast. If you scroll down his logos, IsoSpace was what I worked on, a secure collaboration platform, like webex only better. At least I don't have to work for Cisco today.

I just clicked through his ["And I'm the Dad"](http://andimthedad.com/), his blog, for which he is "Dad and founder." I guess that happened later. My card would say, "Dad and flounder."

dwijnand commented:
> > @adriaanm @SethTisue for the purposes of semantics, modularization is not a feature, so a feature freeze would not apply.
> 
> I tried that argument, it didn't work on Mr. Adriaan Freeze, aka the Snow Miser:

Given writing code is introducing bugs, removing code is removing bugs.  So it would be a bug fix, which are always very welcome. :trollface:

SethTisue commented:
fresh Jenkins run: 
https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-windows/772/ (404 til Jenkins awakens from his thousand year slumber)

SethTisue commented:
Jenkins is green

the only unresolved thing I can find above is Lukas's comment at
https://github.com/scala/scala/pull/7362#discussion_r228912473, @som-snytt is there anything further to be said there...?

som-snytt committed:
Restore syncvar

SethTisue commented:
one more Jenkins run :
https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-windows/788/
